File: day9/day9.cpp
#include <fstream>
#include <iostream>

#include <fmt/format.h>

using namespace std;

static int cur;
static size_t discarded;

static void scan_next(istream& input)
{
	enum {OUT, GARBAGE, IGNORE} state = OUT;
	while ((cur = input.get()) != EOF)
	{
		if (state == OUT)
		{
			if (cur == '<')
			{
				state = GARBAGE;
			}
			else if (cur == '{' || cur == '}')
			{
				break;
			}
		}
		else if (state == GARBAGE)
		{
			if (cur == '!')
			{
				state = IGNORE;
			}
			else if (cur == '>')
			{
				state = OUT;
			}
			else
			{
				// count the discarded
				discarded++;
			}
		}
		else if (state == IGNORE)
		{
			state = GARBAGE;
		}
		else
		{
			cerr << "Unknown state" << endl;
			abort();
		}
	}
}

static void match(istream& in, int val)
{
	if (val != cur)
	{
		cerr << fmt::format("Expected {}, found {} instead.", val, cur) << endl;
		abort();
	}
	scan_next(in);
}

static int group(istream& in, int points)
{
	int r = 0;
	while (cur == '{')
	{
		r += points;
		scan_next(in);
		r += group(in, points+1);
		match(in, '}');
	}
	return r;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}

	scan_next(input);
	int points = group(input, 1);
	match(input, EOF);
	input.close();

	cout << fmt::format("Part1: {}", points) << endl
	     << fmt::format("Part2: {}", discarded) << endl;
	return 0;
}

File: day9/day9.py
#!/usr/bin/env python3

import sys

cur = 0
discarded = 0

def scan_next(f):
    global cur, discarded
    state = 0
    while True:
        cur = f.read(1)
        if not cur:
            break

        if state == 0:
            if cur == '<':
                state = 1
            elif cur in "{}":
                break
            else:
                pass
        elif state == 1:
            if cur == "!":
                state = 2
            elif cur == ">":
                state = 0
            else:
                discarded += 1
        elif state == 2:
            state = 1
        else:
            print("Error: unknown state", file=sys.stderr)
            sys.exit(1)

def match(f, val):
    global cur
    if val != cur:
        print("Error: expected {}, found {}".format(val, cur), file=sys.stderr)
        sys.exit(1)
    scan_next(f)

def group(f, points):
    global cur
    p = 0
    while cur == "{":
        p += points
        scan_next(f)
        p += group(f, points+1)
        match(f, "}")
    return p

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            scan_next(f)
            points = group(f, 1)
            match(f, "")
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", points)
    print("Part2:", discarded)
